### PR TITLE
etcd-runner: remove mutex on validate() and release() in global.go

### LINKED ISSUE #7891: etcd-runner: fails on etcd-tester (election, no progress after 1 minute)
```
2017-05-05 20:38:04.251694 I | etcd-tester: [round#0 case#2] recovered failure
2017-05-05 20:38:04.251726 I | etcd-tester: [round#0 case#2] pausing the stressers...
2017-05-05 20:38:04.252160 I | etcd-tester: lease stresser "10.240.0.2:2379" is closed
2017-05-05 20:38:04.252322 I | etcd-tester: keyStresser "10.240.0.2:2379" is closed
2017-05-05 20:38:04.254374 I | etcd-tester: lease stresser "10.240.0.4:2379" is closed
2017-05-05 20:38:04.254689 I | etcd-tester: lease stresser "10.240.0.6:2379" is closed
2017-05-05 20:38:04.254839 I | etcd-tester: lease stresser "10.240.0.3:2379" is closed
2017-05-05 20:38:04.255294 I | etcd-tester: lease stresser "10.240.0.5:2379" is closed
2017-05-05 20:38:04.255727 I | etcd-tester: keyStresser "10.240.0.4:2379" is closed
2017-05-05 20:38:04.255819 I | etcd-tester: keyStresser "10.240.0.6:2379" is closed
2017-05-05 20:38:04.255859 I | etcd-tester: keyStresser "10.240.0.5:2379" is closed
2017-05-05 20:38:04.255906 I | etcd-tester: keyStresser "10.240.0.3:2379" is closed
2017-05-05 20:38:04.255930 I | etcd-tester: [round#0 case#2] paused stressers
2017-05-05 20:38:04.255938 I | etcd-tester: [round#0 case#2] wait until cluster is healthy
2017-05-05 20:38:04.283546 I | etcd-tester: [round#0 case#2] cluster is healthy
2017-05-05 20:38:04.283572 I | etcd-tester: [round#0 case#2] checking consistency and invariant of cluster
2017-05-05 20:38:10.325973 I | etcd-tester: [round#0 case#2] (/home/gyuho/go/bin/etcd-runner [election 1494016619491935753 --dial-timeout=10s --endpoints 10.240.0.4:2379 --total-client-connections=10 --rounds=0 --req-rate 100]) stderr 20:37:59.977209 no progress after 1 minute!
panic: no progress after 1 minute!

goroutine 1 [running]:
log.Panic(0xc420339ad8, 0x1, 0x1)
	/usr/local/go/src/log/log.go:322 +0xc0
```

> `etcd-tester --enable-pprof --etcd-runner=$etcd-runner --stresser=keys,lease,election-runner,watch-runner,lock-racer-runner,lease-runner --agent-endpoints="$(echo $AGENT_RPC_ENDPOINTS)"`

### PR DESCRIPTION
**Pull request by fanminshi**
election runner can deadlock in atomic release().

suppose election runner has two clients A and B.
if A is a leader and B is a follower, B obtains lock
for release() and waits for A to close(nextc) which signal
next round is ready. However, A can only close(nextc) if it
obtains lock for release(); hence deadlock.

this pr removes atomicity of validate() and release() in global.go
and gives the responsibility of locking to each runner.

FIXES #7891

### CODE REVIEW

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go [hunk 1]
```diff
@@ -69,12 +72,16 @@ func runRacerFunc(cmd *cobra.Command, args []string) {
 		m := concurrency.NewMutex(s, racers)
 		rcs[i].acquire = func() error { return m.Lock(ctx) }
 		rcs[i].validate = func() error {
+			mu.Lock()
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: how is this functionally different from what's in the code now?

The current code:
```go
mu.Lock()
f()
mu.Unlock()
```

does the same thing as this new code:
```go
func f() {
    mu.Lock()
    defer mu.Lock()
    ...
}
```
(panics not withstanding)

Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: OK nevermind the mutex isn't being applied to the election. Is this the right fix, though? Why is a follower `acquire`ing the election-- shouldn't acquire mean exclusive ownership?

Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: Maybe just tear our the entire doRounds acquire/release thing for elections. There's no way to sanely reason about this.

**Comment by heyitsanthony** [comment]
Tried this with `-race`?
```
==================
WARNING: DATA RACE
Write at 0x00c4201c2460 by goroutine 154:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func2()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:93 +0x392
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds.func1()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:72 +0x1f2

Previous write at 0x00c4201c2460 by goroutine 161:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func2.1()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:83 +0x165

Goroutine 154 (running) created at:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:86 +0x23e
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:142 +0xbec
  github.com/spf13/cobra.(*Command).execute()
      /Users/anthony/go/src/github.com/spf13/cobra/command.go:572 +0x801
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/anthony/go/src/github.com/spf13/cobra/command.go:662 +0x458
  github.com/spf13/cobra.(*Command).Execute()
      /Users/anthony/go/src/github.com/spf13/cobra/command.go:618 +0x38
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.Start()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/root.go:66 +0x107
  main.main()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/main.go:21 +0x2f

Goroutine 161 (finished) created at:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func2()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:90 +0x161
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds.func1()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:72 +0x1f2
==================
```

**Comment by fanminshi** [comment]
@heyitsanthony good catch on the data race. let me see what's going on.

File: tools/functional-tester/etcd-runner/command/lock_racer_command.go (hunk 1 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: `The follower acquires the election but the doRounds code is assuming exclusive ownership and that's causing a deadlock.`

Follower can't win the election if it is a follower. Only one client can win the election. The deadlock is caused when the follower block on [`rcNextc`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L135) when it obtained [`mu.Lock()`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/global.go#L86) before leader does.  However, if leader obtains  [`mu.Lock()`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/global.go#L86) before follower does, then everything is fine.

remove the lock fixes the above issue.

Comment (heyitsanthony) [inline tools/functional-tester/etcd-runner/command/lock_racer_command.go]: I was talking about whether the doRounds loop is the right abstraction here. I know how elections work.

**Comment by fanminshi** [comment]
The race is on var [`observedLeader `](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L560 which is a local variable for the each client. It is not caused by this pr.

I'll have another pr to fix that.

**Comment by heyitsanthony** [comment]
@fanminshi the race should be fixed here

**Comment by fanminshi** [comment]
@heyitsanthony alright will fix the race https://github.com/coreos/etcd/issues/7903 in this pr too.

**Comment by fanminshi** [comment]
fixed #7903.

**Comment by heyitsanthony** [comment]
```
==================
WARNING: DATA RACE
Read at 0x00c420078118 by goroutine 161:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func1()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:60 +0x67
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func3()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:115 +0x132
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds.func1()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:75 +0x1ff

Previous write at 0x00c420078118 by goroutine 151:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc.func4()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:133 +0x3ea
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds.func1()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:83 +0x314

Goroutine 161 (running) created at:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:86 +0x23e
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:142 +0xbec
  github.com/spf13/cobra.(*Command).execute()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:572 +0x801
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:662 +0x458
  github.com/spf13/cobra.(*Command).Execute()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:618 +0x38
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.Start()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/root.go:66 +0x107
  main.main()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/main.go:21 +0x2f

Goroutine 151 (running) created at:
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.doRounds()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/global.go:86 +0x23e
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.runElectionFunc()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/election_command.go:142 +0xbec
  github.com/spf13/cobra.(*Command).execute()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:572 +0x801
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:662 +0x458
  github.com/spf13/cobra.(*Command).Execute()
      /Users/anthony/hm/src/github.com/spf13/cobra/command.go:618 +0x38
  github.com/coreos/etcd/tools/functional-tester/etcd-runner/command.Start()
      /Users/anthony/hm/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/command/root.go:66 +0x107
  main.main()
      /Users/anthony/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-runner/main.go:21 +0x2f
==================
```

**Comment by fanminshi** [comment]
I am unsure how the above race can happen. 

 [`setRcNextc()`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L115) races against [`nextc = make(chan struct{})`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L133)

The validateWaiters loop ensures that all followers have been validated in which [`setRcNextc()`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L115) is executed.
```
for validateWaiters > 0 {
	select {
	case <-validatec:
		validateWaiters--
	default:
		return fmt.Errorf("waiting on followers")
	}
}
```
https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go

before leader  sets `nextc`
```go
if observedLeader == v {
	close(nextc)
	nextc = make(chan struct{}) // race here
} 
```
https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L133

**Comment by fanminshi** [comment]
Fixed race on `nextc`

```go
if observedLeader == v {
	close(nextc)
	nextc = make(chan struct{}) // race here
} 
```
leader `close(nextc)` before `nextc = make(chan struct{}` can cause follower to start next around and execute [`setRcNextc()`](https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-runner/command/election_command.go#L115) at same time the leader from executing `nextc = make(chan struct{}`.

**Pull request merged**